Fix for excess IP release race condition | handshake between agent and operator

hemanthmalla opened the pull request:
Forward port of https://github.com/DataDog/cilium/pull/24

Currently there's a 15 sec delay in updating ciliumnode (CN) CRD after IP allocation to a pod, meanwhile the operator can determine that a node has excess IPs and release the IP causing pod connectivity issues.

This PR introduces a handshake between the operator and agent before releasing an excess IP to avoid the race condition. A new operator flag `excess-ip-release-delay` is added to control how long operator should wait before considering an IP for release (defaults to 180 secs). This is done to better handle IP reuse during rollouts. Operator and agent use a new map in cilium node status `.status.ipam.release-ips` to exchange information during the handshake.

Fixes: #13412

![handshake_upstream](https://user-images.githubusercontent.com/3775612/142660918-132aa9e7-d59d-466b-92e1-86492ffef213.png)

The following alternative options were considered : 

* We could force a CN write to go through before allocation but it can result in too many writes from the agent.

* The operator picks the ENI with the most free IPs to release excess IPs from, this is done so that the ENI could potentially be released in the future. We could move the excess detection logic to the agent, but this involves calling into cloud provider specific implementation, which AFAIU the agent IPAM implementation does not.


```release-note
Fix for excess IP release race condition. New operator flag excess-ip-release-delay is introduced to control waiting period before marking an IP for release.
```
Note : https://github.com/DataDog/cilium/pull/24 already received some valuable feedback from @aanm @gandro and @christarazi

hemanthmalla committed:
Adding handshake between agent and operator before releasing IPs.

Currently there's a potential 15 sec delay in updating ciliumnode CRD after IP
allocation to a pod, meanwhile the operator can determine that a node has
excess IPs and release the IP causing pod connectivity issues.

A new operator flag `excess-ip-release-delay` is added to control how long
operator should wait before considering an IP for release(defaults to 180 secs).
This is done to better handle IP reuse during rollouts. Operator and agent use
a new map in cilium node status .status.ipam.release-ips to exchange information
during the handshake.

Fixes: #13412

Signed-off-by: Hemanth Malla <hemanth.malla@datadoghq.com>

gandro commented on pkg/ipam/crd.go:
```diff
@@ -323,6 +323,54 @@ func (n *nodeStore) updateLocalNodeResource(node *ciliumv2.CiliumNode) {
+				ipFamily = IPv4
+			} else {
+				ipFamily = IPv6
+			}
```
tiny nit (feel free to ignore): 
```suggestion
			ipFamily = DeriveFamiliy(ipAddr)
```

hemanthmalla replied:
Updated this.

gandro commented on pkg/ipam/crd.go:
```diff
@@ -389,6 +437,12 @@ func (n *nodeStore) allocate(ip net.IP) (*ipamTypes.AllocationIP, error) {
+		if status == ipamOption.IPAMMarkForRelease || status == ipamOption.IPAMReadyForRelease {
+			return nil, fmt.Errorf("IP not available, marked or ready for release")
+		}
+	}
```
Sorry I didn't realize this last time, but a check like this also needs to be added to `nodeStore.allocateNext()`. Otherwise the IP will be handed out, as IPs are usually allocated via `allocateNext()`.

i.e. adding a check like this:
```go
	if status, ok := n.ownNode.Status.IPAM.ReleaseIPs[ip]; ok {
		if status == ipamOption.IPAMMarkForRelease || status == ipamOption.IPAMReadyForRelease {
			continue // IP not available
		}
	}
```

to the iteration here:
https://github.com/cilium/cilium/blob/a10a784bfba4eae0651a7387de95216fb3ec4223/pkg/ipam/crd.go#L465

In addition, I think we also need to add a similar check to `hasMinimumIPsInPool`.

hemanthmalla replied:
Added checks to allocateNext() and updated `hasMinimumIPsInPool` to factor in IPs currently in handshake.

gandro commented on pkg/ipam/types/types.go:
```diff
@@ -114,6 +114,15 @@ type IPAMStatus struct {
+	// * do-not-release     : IP already in use / not owned by the node. Set by agent
+	//
+	// +optional
+	ReleaseIPs map[string]string `json:"release-ips,omitempty"`
```
nit: You could introduce a newtype with K8s validation requirements here. For example:

```go
// +kubebuilder:validation:Enum=marked-for-release;ready-for-release;do-not-release
type IPReleaseStatus string
```

And then change this type to:

```suggestion
	ReleaseIPs map[string]IPReleaseStatus `json:"release-ips,omitempty"`
```

hemanthmalla replied:
Added the new type and validations. Had to update some tests to convert from `IPReleaseStatus ` to string

gandro reviewed (changes requested):
Awesome, thanks a ton for addressing this long-standing issue.

I think we need to ensure that `allocateNext` doesn't hand out IPs which are marked for release too (see inline comment), the rest are smaller nits.

hemanthmalla committed:
Fix for CN update order issue by adding new handshake state "released"

After the handshake is complete and the operator is done releasing an IP,
CiliumNode status (release-ips) and spec (pool) are updated in two
consecutive requests. There's a tiny window between the two updates where
the entry is removed from .status.ipam.release-ips but the IP is still
present in spec.ipam.pool. It was possible that the IP could be allocated
between these requests.

This commit introduces a new state called released to deal with this.
Now agent removes the entry from release-ips only when the IP was
removed from .spec.ipam.pool as well.

Signed-off-by: Hemanth Malla <hemanth.malla@datadoghq.com>

gandro reviewed (commented):
The implementation looks good to me now! Thanks al lot!

There seems to be a unit test failure that needs to be fixed. Let me know if you need any help with that.

Edit: https://app.travis-ci.com/github/cilium/cilium/jobs/549921704

gandro reviewed (approved):
Discussed offline. Looks like the previous Travis failure was actually unrelated and a case of the flaky unit tests https://github.com/cilium/cilium/issues/11560

Looks good to me now! :rocket:

christarazi commented on operator/provider_aws_flags.go:
```diff
@@ -29,6 +29,9 @@ func init() {
 
+	flags.Int(operatorOption.ExcessIPReleaseDelay, 180, "Number of seconds operator would wait before it releases an IP previously marked as excess. Defaults to 180")
+	option.BindEnv(operatorOption.ExcessIPReleaseDelay)
+
```
I think the `flags` package will actually already show the default value in the text string, so no need to add it yourself.

hemanthmalla replied:
Removed the default part.

christarazi commented on pkg/ipam/crd.go:
```diff
@@ -330,6 +330,54 @@ func (n *nodeStore) updateLocalNodeResource(node *ciliumv2.CiliumNode) {
+		}
+		allocator.mutex.Unlock()
+		releaseUpstreamSyncNeeded = true
+	}
```
Would it make sense to move this code to a isolated function and have it return `releaseUpstreamSyncNeeded`?

christarazi commented on pkg/ipam/node.go:
```diff
@@ -567,39 +569,124 @@ func (n *Node) determineMaintenanceAction() (*maintenanceAction, error) {
+			// n.determineMaintenanceAction() only returns the IPs on the interface with maximum number of IPs that
+			// can be freed up. If the selected interface changes or if this IP is not excess anymore, remove entry
+			// from local map and cilium node status if n.ipReleaseStatus is already updated.
+			// from n.ipReleaseStatus
```
Is the second to last sentence supossed to end or is `from n.ipReleaseStatus` part of it?

hemanthmalla replied:
This was a stale comment ( probably from applying patch). Fixed this.

christarazi commented on pkg/ipam/node.go:
```diff
@@ -567,39 +569,124 @@ func (n *Node) determineMaintenanceAction() (*maintenanceAction, error) {
+	}
+	// Update cilium node CRD with IPs that need to be marked for release
+	for _, ip := range ipsToMark {
+		scopedLog.WithFields(logrus.Fields{"ip": ip}).Debug("Marking IP for release")
```
logfields.IPAddr

christarazi commented on pkg/ipam/node.go:
```diff
@@ -567,39 +569,124 @@ func (n *Node) determineMaintenanceAction() (*maintenanceAction, error) {
+	if len(ipsToRelease) > 0 {
+		a.release.IPsToRelease = ipsToRelease
+		scopedLog = scopedLog.WithFields(logrus.Fields{
+			"available":         n.stats.AvailableIPs,
+			"used":              n.stats.UsedIPs,
+			"excess":            n.stats.ExcessIPs,
+			"excessIps":         a.release.IPsToRelease,
+			"releasing":         ipsToRelease,
+			"selectedInterface": a.release.InterfaceID,
+			"selectedPoolID":    a.release.PoolID,
```
Ideally, since these are a part of an INFO level log, they should be inside the logfields package. However, there are other parts of the code that also suffer the same thing, so let's just followup on this later, doesn't have to be part of this PR.

christarazi commented on pkg/ipam/crd.go:
```diff
@@ -346,11 +349,7 @@ func (n *nodeStore) updateLocalNodeResource(node *ciliumv2.CiliumNode) {
-			} else {
-				ipFamily = IPv6
-			}
+			ipFamily = DeriveFamily(ipAddr)
```
This could be part of the first commit, no?

hemanthmalla replied:
The first and second commits are now squashed.

christarazi commented on pkg/ipam/crd.go:
```diff
@@ -337,12 +337,20 @@ func (n *nodeStore) updateLocalNodeResource(node *ciliumv2.CiliumNode) {
 	releaseUpstreamSyncNeeded := false
 	// ACK or NACK IPs marked for release by the operator
 	for ip, status := range n.ownNode.Status.IPAM.ReleaseIPs {
-		if status != ipamOption.IPAMMarkForRelease || n.ownNode.Spec.IPAM.Pool == nil {
```
nit on commit msg title: We could probably just shorten it to 

```
Fix CN update order by adding new `released` IPAM state
```

hemanthmalla replied:
Updated the commit message.

christarazi reviewed (approved):
Thanks for forward porting! The changes LGTM overall. I am marking my review as approved, but there are suggestions that I'd like to see followed up on in a later PR, so that we can get this PR in. Feel free to address my comments as you see fit in this PR however. After this is merged, we can create a tracking issue and put all the comments that need following up on in there. Please use "resolve conversion" for comments that you've already addressed in this PR.

Note, for any suggestions related to commits, could you address those in this PR as those can't be changed in a followup?

wu0407 commented on pkg/aws/eni/node.go:
```diff
@@ -119,9 +120,17 @@ func (n *Node) PrepareIPRelease(excessIPs int, scopedLog *logrus.Entry) *ipam.Re
+	for k := range n.enis {
+		eniIds = append(eniIds, k)
+	}
+	sort.Strings(eniIds)
```
Same iterate order not ensure selecting the same ENI to release IPs in multi `PrepareIPRelease` called. what is use case ? @hemanthmalla

wu0407 replied:
Because select the ENI with the most addresses available for release  in every call `PrepareIPRelease`
